circus-train-gcp/src/test/java/com/hotels/bdp/circustrain/gcp/GCPCredentialCopierTest.java

```diff
@@ -16,59 +16,128 @@
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.fs.Path;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
 import org.mockito.Mock;
 import org.powermock.core.classloader.annotations.PrepareForTest;
 import org.powermock.modules.junit4.PowerMockRunner;
 
 import fm.last.commons.test.file.TemporaryFolder;
 
 import com.hotels.bdp.circustrain.api.CircusTrainException;
+import com.hotels.bdp.circustrain.gcp.context.GCPSecurity;
 
 @RunWith(PowerMockRunner.class)
-@PrepareForTest({ FileSystem.class })
+@PrepareForTest({ FileSystem.class, FileUtils.class })
 public class GCPCredentialCopierTest {
 
-  private @Rule TemporaryFolder temporaryFolder = new TemporaryFolder();
-  private @Mock FileSystem fs;
-  private Configuration conf = new Configuration();
+  private static final String DISTRIBUTED_CACHE_PROPERTY = "mapreduce.job.cache.files";
+  private static final String SYMLINK_FLAG = "#";
+
+  private @Rule final TemporaryFolder temporaryFolder = new TemporaryFolder();
+  private @Mock FileSystem fileSystem;
+  private final Configuration conf = new Configuration();
+  private final GCPSecurity gcpSecurity = new GCPSecurity();
+  private final String distributedFileSystem = "hdfs:/tmp/circus-train-gcp/workdir/";
+  private final String defaultDistributedFileSystemRoot = "hdfs:/tmp/ct-gcp-";
+
+  private String credentialsFilePath = "/test.json";
+  private final String credentialsFileRelativePath = "../../../../../.." + credentialsFilePath;
+
+  private void setGcpSecurity(String credentialProvider, String distributedFileSystemWorkingDirectory) {
+    gcpSecurity.setCredentialProvider(credentialProvider);
+    gcpSecurity.setDistributedFileSystemWorkingDirectory(distributedFileSystemWorkingDirectory);
+  }
 
   @Test(expected = IllegalArgumentException.class)
   public void credentialProviderNotSetThrowsException() throws Exception {
-    doNothing().when(fs).copyFromLocalFile(any(Path.class), any(Path.class));
-    doReturn(false).when(fs).exists(any(Path.class));
-    String credentialProvider = null;
-    GCPCredentialCopier copier = new GCPCredentialCopier(fs, conf, credentialProvider);
+    doNothing().when(fileSystem).copyFromLocalFile(any(Path.class), any(Path.class));
+    doReturn(false).when(fileSystem).exists(any(Path.class));
+    new GCPCredentialCopier(fileSystem, conf, gcpSecurity);
   }
 
   @Test
-  public void copyCredentials() throws Exception {
-    doNothing().when(fs).copyFromLocalFile(any(Path.class), any(Path.class));
-    String credentialProvider = "test.json";
-    temporaryFolder.newFile(credentialProvider);
-    credentialProvider = temporaryFolder.getRoot() + "/" + credentialProvider;
-    GCPCredentialCopier copier = new GCPCredentialCopier(fs, conf, credentialProvider);
+  public void copyCredentialsWithCredentialProviderSupplied() throws Exception {
```

Comment:
Remove the `shouldGetRelativePath` test method and combine the assert you do there in one test like this:


```
 @Test
  public void copyCredentialsWithCredentialProviderSupplied() throws Exception {
    setGcpSecurity(credentialsFilePath, null);
    GCPCredentialCopier copier = new GCPCredentialCopier(fileSystem, conf, gcpSecurity);
    copier.copyCredentials();

    verify(fileSystem).copyFromLocalFile(eq(new Path(credentialsFileRelativePath)), any(Path.class));
    assertNotNull(conf.get(DISTRIBUTED_CACHE_PROPERTY));
    assertThat(conf.get(DISTRIBUTED_CACHE_PROPERTY), endsWith(SYMLINK_FLAG + credentialsFileRelativePath));
  }
  }

```

imo, it more clear that you copy does two things, copy to hdfs and make the cache entry.

Then add another test where the credentialsFilePath is configured as a relative file path.

A further improvement would be like @teabot  suggested to add a factory for the `hdfsGsCredential` paths. E.g. `HdfsGsCredentialPathGenerator`, has one method `newPath()`.  You can add the random stuff in there and mock the class in this test to do better verifications.
You can't assert the deleteOnExit very well atm, this generator class will help with that.

Reply:
I added factories for building some paths, so tests can be more separate from each other and some elements can be mocked, which makes testing a bit easier. I'm going to push the changes as soon as I finish testing the build.